server/permalinks.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"context"
+	"encoding/hex"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/v25/github"
+)
+
+const maxPermalinkReplacements = 10
+const permalinkReqTimeout = 2 * time.Second
+const maxPreviewLines = 10
+
+// replacement holds necessary info to replace github permalinks
+// in messages with a code preview block
+type replacement struct {
+	index      int               // index of the permalink in the string
+	word       string            // the permalink
+	captureMap map[string]string // named regex capture group of the link
+}
+
+// getReplacements returns the permalink replacements that needs to be performed
+// on a message.
+func (p *Plugin) getReplacements(msg string) []replacement {
+	// find the permalinks from the msg using a regex
+	matches := p.githubPermalinkRegex.FindAllStringSubmatch(msg, -1)
+	indices := p.githubPermalinkRegex.FindAllStringIndex(msg, -1)
+	replacements := make([]replacement, 0, maxPermalinkReplacements)
+	for i, m := range matches {
+		// have a limit on the no. of replacements to do
+		if i > maxPermalinkReplacements {
+			break
+		}
+		word := m[0]
+		index := indices[i][0]
+		r := replacement{
+			index: index,
+			word:  word,
+		}
+		// ignore if the word is inside a link
+		if isInsideLink(msg, index) {
+			continue
+		}
+		// populate the captureMap with the extracted groups of the regex
+		r.captureMap = make(map[string]string)
+		for j, name := range p.githubPermalinkRegex.SubexpNames() {
+			if j == 0 {
+				continue
+			}
+			r.captureMap[name] = m[j]
+		}
+		replacements = append(replacements, r)
+	}
+	return replacements
+}
+
+// makeReplacements perform the given replacements on the msg and returns
+// the new msg.
+func (p *Plugin) makeReplacements(msg string, replacements []replacement) string {
+	for _, r := range replacements {
+		// quick bailout if the commit hash is not proper
+		if _, err := hex.DecodeString(r.captureMap["commit"]); err != nil {
+			p.API.LogError("bad git commit hash in permalink", "error", err.Error(), "hash", r.captureMap["commit"])
+			continue
+		}
+
+		// get the file contents
+		opts := github.RepositoryContentGetOptions{
+			Ref: r.captureMap["commit"],
+		}
+		// TODO: make all of these requests concurrently.
+		reqctx, cancel := context.WithTimeout(context.Background(), permalinkReqTimeout)
+		fileContent, _, _, err := p.githubClient.Repositories.GetContents(reqctx,
+			r.captureMap["user"], r.captureMap["repo"], r.captureMap["path"], &opts)
+		if err != nil {
+			p.API.LogError("error while fetching file contents", "error", err.Error(), "path", r.captureMap["path"])
+			cancel()
+			continue
+		}
+		cancel()
```

Comment:
Ah no, we need this. The context needs to be cancelled after the call is done. We cannot use `defer` because it won't get cancelled until the function returns which is something we don't want. Then we need to declare a closure and that hurts readability.

Reply:
Shouldn't this be updated to `defer cancel()`?

Reply:
```
// TODO: make all of these requests concurrently.
		reqctx, cancel := context.WithTimeout(context.Background(), permalinkReqTimeout)
		fileContent, _, _, err := ghClient.Repositories.GetContents(reqctx, r.permalinkInfo.user, r.permalinkInfo.repo, r.permalinkInfo.path, &opts)
		defer cancel()
```

Reply:
Ah ok, it was not recommended to use `defer` in loops as it has performance implications. But now with open-coded defers in tip, this should be fine. Will change.

Reply:
@agnivade hmmm fair point. Yeah I see folks are recommending to close at the end of the loop rather than doing `defer cancel()` Maybe we could do that instead?